hal/halx86/acpi/madt.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Source File for MADT parses 
+ * PROGRAMMERS: Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ */
+
+/* INCLUDES *******************************************************************/
+
+#include <hal.h>
+#include <acpi.h>
+/* ACPI_BIOS_ERROR defined in acoutput.h and bugcodes.h */
+#undef ACPI_BIOS_ERROR
+#include <smp.h>
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS ********************************************************************/
+
+PROCESSOR_IDENTITY HalpStaticProcessorIdentity[MAX_CPUS] = {{0}};
+PPROCESSOR_IDENTITY HalpProcessorIdentity = NULL;
+HALP_APIC_INFO_TABLE HalpApicInfoTable;
+UINT32 PhysicalProcessorCount = 0;
+ACPI_TABLE_MADT *MadtTable;
+ACPI_SUBTABLE_HEADER *AcpiHeader;
+ACPI_MADT_LOCAL_APIC *LocalApic;
+
+/* FUNCTIONS ******************************************************************/
+
+VOID
+HalpParseApicTables(IN PLOADER_PARAMETER_BLOCK LoaderBlock)
+{
+    ULONG_PTR TableEnd;
+    ULONG Count;
+    ULONG ValidProcessorCount;
+
+    /* We only support legacy APIC for now, this will be updated in the future */
+    HalpApicInfoTable.ApicMode = 0x10;
+    MadtTable = HalAcpiGetTable(LoaderBlock, 'CIPA');
+
+    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
+    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
+    TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+
+    HalpProcessorIdentity = HalpStaticProcessorIdentity;
+
+    while ((ULONG_PTR)AcpiHeader <= TableEnd)
+    {
+        LocalApic = (ACPI_MADT_LOCAL_APIC*)AcpiHeader;
+        if (LocalApic->Header.Type == ACPI_MADT_TYPE_LOCAL_APIC && 
+            LocalApic->Header.Length == sizeof(ACPI_MADT_LOCAL_APIC))
+        {
+            PhysicalProcessorCount++;
+        }
+
+        if (!AcpiHeader->Length)
+        {
+            break;
+        }
+
+        AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + AcpiHeader->Length);
+    }
+
+    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
```

Comment:
Is this going to be kind of a re-parsing of the table above? If so, could both loops be merged together somehow?
I note for example, that the loop above always updates the `AcpiHeader` pointer with its "old" Length, while the loop below does different ways (see the `else` case); therefore I'm not sure whether they all kind of parse the same way.

Reply:
I could give an attempt at merging it, they serve a slightly different purpose but generally it should be possible to do it in one go. However as we expand our APIC this loop will expand as it goes through the tables. Generally the parse where I get the IDs won't always happen that way since that's for older APIC versions. So there will be situations where we're parsing the x2APIC ID table but using the processor count from parsing the Local APIC Table and I personally just wanted to differentiate the two _visually_